### PR TITLE
K8SPSMDB-713: Run PBM resync after physical restore

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPSMDB-713](https://jira.percona.com/browse/K8SPSMDB-713) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

**CHANGE DESCRIPTION**
---
**Problem:**
PBM needs a storage resync after physical restore.

**Solution:**

After physical restore finishes, restore controller annotates the `PerconaServerMongoDB` object with `percona.com/resync-pbm`. psmdb controller sees the annotation, removes it and starts resync. **It doesn't wait for resync to finish, just start.**

**CHECKLIST**
---
**Jira**
- [ ] Is the Jira ticket created and referenced properly?
- [ ] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [ ] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [ ] Are E2E tests passing?
- [ ] Are unit tests passing?
- [ ] Are linting tests passing?
- [ ] Is an E2E test/test case added for the new feature/change?
- [ ] Are unit tests added where appropriate?
- [ ] Are OpenShift compare files changed for E2E tests (`compare/*-oc.yml`)?

**Config/Logging/Testability**
- [ ] Are all needed new/changed options added to default YAML files?
- [ ] Are the manifests (crd/bundle) regenerated if needed?
- [ ] Did we add proper logging messages for operator actions?
- [ ] Did we ensure compatibility with the previous version or cluster upgrade process?
- [ ] Does the change support oldest and newest supported MongoDB version?
- [ ] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/upgrade-consistency.log) |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/cross-site-sharded.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/storage.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/one-pod.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/demand-backup.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/self-healing-chaos.log) |
| ignore-labels-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/ignore-labels-annotations.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/scaling.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/monitoring-2-0.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/operator-self-healing-chaos.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/security-context.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/scheduled-backup.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/arbiter.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/rs-shard-migration.log) |
| expose-sharded | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/expose-sharded.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/service-per-pod.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/demand-backup-sharded.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/liveness.log) |
| demand-backup-physical | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/demand-backup-physical.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/smart-update.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/version-service.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/upgrade.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/users.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/data-sharded.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/upgrade-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/demand-backup-eks-credentials.log) |
| data-at-rest-encryption | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/data-at-rest-encryption.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/pitr.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/e1b5036a/pitr-sharded.log) |
| We run 32 out of 33|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/1151/commits/e1b5036a8c58ca9d10f0feb97b49abffcb144e89
image: `perconalab/percona-server-mongodb-operator:PR-1151-e1b5036a`

File: pkg/controller/perconaservermongodbrestore/physical.go [hunk 1]
```diff
@@ -233,6 +233,23 @@ func (r *ReconcilePerconaServerMongoDBRestore) reconcilePhysicalRestore(ctx cont
...
+
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			c := &psmdbv1.PerconaServerMongoDB{}
+			err := r.client.Get(ctx, types.NamespacedName{Name: cluster.Name, Namespace: cluster.Namespace}, c)
+			if err != nil {
+				return err
+			}
+
+			orig := c.DeepCopy()
+			c.Annotations[psmdbv1.AnnotationResyncPBM] = "true"
```
Comment (inelpandzic) [inline pkg/controller/perconaservermongodbrestore/physical.go:245]: Maybe I'm missing something, but why do we need to do it like this with annotation, could we just perform the resync here directly (asynchronously)?

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/storage.log) |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/cross-site-sharded.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/init-deploy.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/self-healing-chaos.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/limits.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/operator-self-healing-chaos.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/scaling.log) |
| expose-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/expose-sharded.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/security-context.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/upgrade-consistency.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/rs-shard-migration.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/demand-backup.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/scheduled-backup.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/demand-backup-sharded.log) |
| demand-backup-physical | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/demand-backup-physical.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/one-pod.log) |
| ignore-labels-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/ignore-labels-annotations.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/monitoring-2-0.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/arbiter.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/service-per-pod.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/liveness.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/smart-update.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/version-service.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/users.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/data-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/demand-backup-eks-credentials.log) |
| data-at-rest-encryption | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/data-at-rest-encryption.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/upgrade.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/upgrade-sharded.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/pitr.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1151/ba229976/pitr-sharded.log) |
| We run 32 out of 33|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/1151/commits/ba2299767742ca7afe4c1b819c37ad6364fff8d0
image: `perconalab/percona-server-mongodb-operator:PR-1151-ba229976`

File: pkg/controller/perconaservermongodb/backup.go [hunk 2]
```diff
@@ -470,3 +473,53 @@ func (r *ReconcilePerconaServerMongoDB) updatePITR(ctx context.Context, cr *api.
...
+func (r *ReconcilePerconaServerMongoDB) resyncPBMIfNeeded(ctx context.Context, cr *api.PerconaServerMongoDB) error {
+	log := logf.FromContext(ctx)
+
+	if cr.Status.State != api.AppStateReady || !cr.Spec.Backup.Enabled {
+		return nil
+	}
+
+	_, resyncNeeded := cr.Annotations[api.AnnotationResyncPBM]
+	if !resyncNeeded {
+		return nil
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/controller/perconaservermongodb/backup.go:486]: We'll print this log in every reconciliation and 99% of time for nothing. I don't think I agree

File: pkg/controller/perconaservermongodbrestore/physical.go (hunk 1 above)
Comment (egegunes) [reply pkg/controller/perconaservermongodbrestore/physical.go:245]: statefulsets are deleted before this operation. we need to wait until cluster is ready, so we can run resync. that's why I'm doing this with annotation. for doing that here asynchronously, we need to check cluster status until it's ready and then run the resync in a separate goroutine. I think it'll be over complicated.

**Review by hors — APPROVED** [review]

**Pull request merged**